File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.kt
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        var leftIndex = 0
        var maxLength = 0

        val charMap = HashMap<Char, Int>()

        s.forEachIndexed { index, charKey ->
            if (charMap.contains(charKey)) {
                leftIndex = Math.max(charMap.getOrDefault(charKey, 0) + 1, leftIndex)
            }

            maxLength = Math.max(maxLength, index - leftIndex + 1)
            charMap.put(charKey, index)
        } 

        return maxLength
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.rs
use std::collections::HashMap;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut left = 0;
        let mut max_len = 0;
        let mut word_map: HashMap<char, usize> = HashMap::new();

        let mut s_list: Vec<char> = s.chars().collect();


        for (char_index, char_item) in s_list.iter().enumerate() {
            if let Some(old_index) = word_map.insert(*char_item, char_index) {
                left = left.max(old_index + 1);
            }

            let cur_len = char_index - left + 1;

            max_len = max_len.max(cur_len);
        }

        max_len as i32
    }
}